src/charm.py

```diff
@@ -2,21 +2,286 @@
 # Copyright 2023 Canonical Ltd.
 # See LICENSE file for licensing details.
 
-"""Charmed operator for the 5G UDR service."""
+"""Charmed operator for the SD-Core's UDR service."""
 
 import logging
+from ipaddress import IPv4Address
+from subprocess import check_output
+from typing import Optional, Union
 
-from ops.charm import CharmBase
+from charms.data_platform_libs.v0.data_interfaces import (  # type: ignore[import]
+    DatabaseCreatedEvent,
+    DatabaseRequires,
+)
+from charms.observability_libs.v1.kubernetes_service_patch import (  # type: ignore[import]  # noqa: E501
+    KubernetesServicePatch,
+)
+from charms.sdcore_nrf.v0.fiveg_nrf import NRFAvailableEvent, NRFRequires  # type: ignore[import]
+from jinja2 import Environment, FileSystemLoader
+from lightkube.models.core_v1 import ServicePort
+from ops.charm import CharmBase, PebbleReadyEvent, RelationJoinedEvent
 from ops.main import main
+from ops.model import ActiveStatus, BlockedStatus, WaitingStatus
+from ops.pebble import Layer, PathError
 
 logger = logging.getLogger(__name__)
 
+BASE_CONFIG_PATH = "/free5gc/config"
+DEFAULT_DATABASE_NAME = "free5gc"
+UDR_CONFIG_FILE_NAME = "udrcfg.yaml"
+UDR_SBI_PORT = 29504
+
 
 class UDROperatorCharm(CharmBase):
     """Charm the service."""
 
     def __init__(self, *args):
         super().__init__(*args)
+        self._container_name = self._service_name = "udr"
+        self._container = self.unit.get_container(self._container_name)
+
+        self._nrf = NRFRequires(charm=self, relation_name="fiveg_nrf")
+
+        self._database = DatabaseRequires(
+            self, relation_name="database", database_name=DEFAULT_DATABASE_NAME
+        )
+
+        self._service_patcher = KubernetesServicePatch(
+            charm=self,
+            ports=[ServicePort(name="sbi", port=UDR_SBI_PORT)],
+        )
+
+        self.framework.observe(self.on.udr_pebble_ready, self._configure_sdcore_udr)
+        self.framework.observe(self.on.database_relation_joined, self._configure_sdcore_udr)
+        self.framework.observe(self._database.on.database_created, self._configure_sdcore_udr)
+        self.framework.observe(self.on.fiveg_nrf_relation_joined, self._configure_sdcore_udr)
+        self.framework.observe(self._nrf.on.nrf_available, self._configure_sdcore_udr)
+
+    def _configure_sdcore_udr(
+        self,
+        event: Union[
+            DatabaseCreatedEvent, NRFAvailableEvent, PebbleReadyEvent, RelationJoinedEvent
+        ],
```

Comment:
```suggestion
        event: EventBase,
```
In cases like this, the only thing we do with the event is potentially deferring it, so `EventBase` encapsulates all the typing validation that we need.

Reply:
Done.